### PR TITLE
Rename libpacemaker-internal to libpacemaker

### PR DESCRIPTION
**Pull request by clumens**
Move the contents of libtransitioner into it, and the process_pe_message and libpengine_fini functions out of it.

### CODE REVIEW

File: daemons/fenced/pacemaker-fenced.c [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
 #include <crm/cib/internal.h>
 #include <crm/pengine/status.h>
-#include <pacemaker-internal.h>
+#include <pacemaker.h>
```
Comment (kgaillot) [inline daemons/fenced/pacemaker-fenced.c]: Now that libpacemaker will have both private and public APIs, we'll need to reserve pacemaker.h for the public API.

Taking a fresh look at naming in light of recent discussions, it looks like we're moving toward an eventual library split of:
* libpacemaker: high-level (tool) functionality (likely including the current libpe_status and libpe_rules)
* libpacemaker-cluster: the current libcrmcluster (which depends on corosync libraries)
* (maybe) libpacemaker-curses: crm_mon's curses functionality (which depends on curses libraries)
* everything else either in one "low-level" library, or split into libpacemaker-common (lowest level, i.e. most of libcrmcommon and libcrmservice), libpacemaker-client (client APIs, i.e. libcib, libstonithd, and liblrmd, and some of libcrmcommon), and libpacemaker-server (the few functions that are only needed by the daemons).

Looking at that, I'm thinking we should abandon the idea of using a unique prefix for each library, and just use pcmk for everything. It's especially difficult since up to four libraries would start with a "c". :-) I see other projects with multiple libraries that don't tie function prefixes to libraries.

It's also questionable whether we should keep using the include/crm subdirectory. Distros tend to install to /usr/include/pacemaker, so currently everything is in /usr/include/pacemaker/crm/*.h. It's an unnecessary level, and would look even weirder if we started calling it "pacemaker" e.g. /usr/include/pacemaker/pacemaker/pacemaker.h! Looking at other common projects, most don't worry about adding another level, so I think we can abandon crm for the new libraries.

I'm open to suggestions for how to name the internal headers. Currently we tend to put "_internal" in the name somewhere (though not consistently). We could continue with that, or keep the pcmki prefix just for this purpose, or extend the double underbar convention to header file names, or something else.

Bottom line, maybe this header should stay named pacemaker-internal.h, or pacemaker_internal.h.

File: lib/Makefile.am [hunk 2]
```diff
@@ -17,11 +17,11 @@
 #
 MAINTAINERCLEANFILES    = Makefile.in
 
-EXTRA_DIST		= pacemaker.pc.in pacemaker-pe_rules.pc.in pacemaker-pe_status.pc.in $(target_LIBS:%=%.in)
+EXTRA_DIST		= pacemaker.pc.in pacemaker-pe_rules.pc.in pacemaker-pe_status.pc.in pacemaker-internal.pc.in $(target_LIBS:%=%.in)
```
Comment (kgaillot) [inline lib/Makefile.am]: Blech, I forgot about the existing pacemaker.pc.in (for libcrmcommon). Since the new library will have both public and private APIs, I don't want to mark it internal -- maybe libpacemaker.pc.in?

File: lib/pacemaker/Makefile.am [hunk 3]
```diff
@@ -10,21 +10,21 @@ include $(top_srcdir)/Makefile.common
...
+libpacemaker_la_LDFLAGS	+= $(LDFLAGS_HARDENED_LIB)
 
-libpacemaker_internal_la_LIBADD	= $(top_builddir)/lib/pengine/libpe_status.la \
+libpacemaker_la_LIBADD	= $(top_builddir)/lib/pengine/libpe_status.la \
 			$(top_builddir)/lib/cib/libcib.la \
 			$(top_builddir)/lib/transition/libtransitioner.la \
 			$(top_builddir)/lib/lrmd/liblrmd.la
 # -L$(top_builddir)/lib/pils -lpils -export-dynamic -module -avoid-version
-libpacemaker_internal_la_SOURCES	= pcmki_sched_allocate.c \
+libpacemaker_la_SOURCES	= pcmki_sched_allocate.c \
```
Comment (kgaillot) [inline lib/pacemaker/Makefile.am]: I'd abandon the pcmki prefix for source code files and just use pcmk

File: daemons/execd/Makefile.am [hunk 4]
```diff
@@ -43,7 +43,7 @@ cts_exec_helper_LDADD	= $(top_builddir)/lib/common/libcrmcommon.la	\
 			  $(top_builddir)/lib/cib/libcib.la		\
 			  $(top_builddir)/lib/services/libcrmservice.la	\
 			  $(top_builddir)/lib/pengine/libpe_status.la	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply daemons/execd/Makefile.am]: Actually, using `libtool` and referencing embedded libtool archives
only (non `*.la` should then be listed last and the order may matter
in this case), and when the respective `*.la` files have properly
tracked dependencies on other `*.la` without circles, this presumably
doesn't play any role here, since

Wouldn't lose a second on this if we are to stay with `libtool` for
the foreseeable future, especially if there are no circular
dependencies amongst the embedded libtool archives.

My understanding is also that listing all the transitive-over-`*.la`
libraries, without having any symbols of these directly used in the
target, is just a bloat given that these dependencies are already
properly tracked per above.

[the remark this answers is not in the document]
Comment (jnpkrn) [reply daemons/execd/Makefile.am]: (e.g. `meson` combined with GNU `ld` seems to mix `--start-group`
and `--end-group` to the build command structure so as to solve the
ordering problem of static linking so being overly considerate here
is rather make-work, IMHO)

[the remark this answers is not in the document]
Comment (clumens) [reply daemons/execd/Makefile.am]: > My understanding is also that listing all the transitive-over-`*.la`
> libraries, without having any symbols of these directly used in the
> target, is just a bloat given that these dependencies are already
> properly tracked per above.

I believe you are right about this.  Perhaps it would be worth doing a commit first to minimize what's listed in _LDADD and _LIBADD everywhere?  #1701 is related to this.

File: daemons/fenced/pacemaker-fenced.c (hunk 1 above)
Comment (jnpkrn) [reply daemons/fenced/pacemaker-fenced.c]: > (maybe) libpacemaker-curses: crm_mon's curses functionality (which
> depends on curses libraries)

This is symptomatic to what I find a bad functional decomposition of
the problem.  More ideally, front-ends would be rather separa(ted,ble}
formatters of a somehow unified data model, see also
https://bugs.clusterlabs.org/show_bug.cgi?id=5376#c3

Especially formatting to the terminal is rather a unique one-off
functionality presumably having just a single client, hence belongs
to such program itself (or a private library if need be, waiting
for other external client wanting to yield 1:1 data-to-terminal
output, which doesn't sound very probable).

File: daemons/execd/Makefile.am (hunk 4 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply daemons/execd/Makefile.am]: That's good to hear. In that case, I'd just try to get the ADDs that are being changed in this PR done reasonably, and not worry about the rest (since this subsumes both libpengine and libtransitioner, #1701 will be covered).

File: daemons/fenced/pacemaker-fenced.c (hunk 1 above)
Comment (kgaillot) [reply daemons/fenced/pacemaker-fenced.c]: Exactly, #1707 is indeed a unified output formatting model with each output type having its own formatter. Of course, curses output requires linking against curses libraries, so we can't put that portion of the code into a common library.

The end goal for the tools, including crm_mon, is to be able to call their functionality as C API. crm_mon's curses capability makes it a special case. We want to have all of crm_mon's capabilities except curses output as an API in libpacemaker, but have crm_mon itself able to register the curses formatter before calling that same API. Whether the curses formatter is in the crm_mon code, a library private to crm_mon, or an internal library that could be used by future tools is an implementation detail that we can address at that time.

Comment (jnpkrn) [reply daemons/fenced/pacemaker-fenced.c]: Ah, was worried that in-program (most appropriate here, I think,
shared library is for reusing, single user = no reusing) hosting
was on the way out along the actual reusable parts.

Btw. maybe would be interesting to put related clusters (sic) of
private-use-only symbols behind lightly used ones and, at the same
time, without any logging (since it would then defeat "singleton"
principle of libqb logging arrangement at least with linker section
involved, I think, impacting efficiency) into a (single) static library
on its own to be linked with _all_ the binary artefacts (will become
a no-op when they don't ask for any such symbol).  This way, we would
keep the dynamic symbol resolution clutter and overhead (also durable
burden with files in locations where publiclz reusable libraries are
expected) minimal. The static size overhead won't be noticable and
pacemaker parts are mostly required to stay in sync anyway.
Code locality would increase (as opposed to private shared libraries),
which is good.  Afterall that's the path modern languages take
(Go, Rust).

Using private shared libraries "because we can and know to do it,
since we have also actual public libraries" is not a very profound
justification to do the same for the private purposes.
Of course, there's also this very basic modules reusing approach,
there's nothing bad about that either, but can be administratively
more clumsy (price paid for a full control).

File: lib/pacemaker/Makefile.am (hunk 3 above)
Comment (clumens) [reply lib/pacemaker/Makefile.am]: If we're using the same prefix for everything, maybe it makes sense to not use a prefix at all and instead just make sure things are in directories to keep them organized.

Comment (kgaillot) [reply lib/pacemaker/Makefile.am]: A pcmk_ prefix could still distinguish library source files from daemon/tool source files. But as long as the daemons have unique prefixes, I suppose it doesn't matter if the libraries have one, as long as we avoid the daemon prefixes (like sched in this case).

The real goal is unique file names across the project, for when using PCMK_trace_files and reading log messages with trace logging turned on (which shows file names). When you have logs like:

pacemaker.log-20180429.gz:Apr 27 21:44:22 [7525] rhel7-1 pacemaker-schedulerd: (     utils.c:559   )   debug: custom_action:	rhel7-1	Fencing_start_0 (cancelled : quorum)

we might as well not log the file name. The 1.1 code base has 8 files named utils.c. :)

Alternatively we could keep the idea of unique prefixes per library, use pcmk_ for libpacemaker, and try to come up with something imaginative when we reorganize the others.

**Comment by clumens** [comment]
Debian-derived systems sure do not like that LDADD patch.

**Comment by clumens** [comment]
It's probably due to this:  https://salsa.debian.org/mckinstry/libtool/blob/debian/master/debian/patches/link_all_deplibs.patch

I'd prefer to just drop the last patch and deal with the linking separately.

**Comment by jnpkrn** [comment]
On 04/04/19 07:20 -0700, Chris Lumens wrote:
> It's probably due to this:
> https://salsa.debian.org/mckinstry/libtool/blob/debian/master/debian/patches/link_all_deplibs.patch

It's not if the tracking is done right.

> I'd prefer to just drop the last patch and deal with the linking
> separately.

The problem here is that libpacemaker_la_LIBADD lacks dependency on
libcrmcommon.la.

(I expressly mentioned "without having any symbols of these directly
used in the target" in the discussion above, apparently crm_xml_add_int
is used directly there.)

-- 
Jan (Poki)

File: daemons/fenced/pacemaker-fenced.c (hunk 1 above)
Comment (kgaillot) [reply daemons/fenced/pacemaker-fenced.c]: libpacemaker isn't solely about reusability; a major goal is to expose the tool functionality as C API so that external projects can use them to create language bindings, API servers, high-level tools, etc.

File: daemons/schedulerd/Makefile.am [hunk 5]
```diff
@@ -27,7 +23,8 @@ endif
 
 pacemaker_schedulerd_CFLAGS	= $(CFLAGS_HARDENED_EXE)
 pacemaker_schedulerd_LDFLAGS	= $(LDFLAGS_HARDENED_EXE)
-pacemaker_schedulerd_LDADD	= $(top_builddir)/lib/cib/libcib.la $(COMMONLIBS)
+pacemaker_schedulerd_LDADD	= $(top_builddir)/lib/pacemaker/libpacemaker.la \
+							  $(top_builddir)/lib/pengine/libpe_status.la
```
Comment (kgaillot) [inline daemons/schedulerd/Makefile.am]: pacemaker-schedulerd also uses libcrmcommon directly

File: tools/Makefile.am [hunk 6]
```diff
@@ -71,46 +71,28 @@ crm_node_LDADD		= $(top_builddir)/lib/cib/libcib.la		\
 
 crm_simulate_SOURCES	= crm_simulate.c
 
-crm_simulate_LDADD	= $(top_builddir)/lib/pengine/libpe_status.la	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la	\
-			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/lrmd/liblrmd.la		\
-			  $(top_builddir)/lib/transition/libtransitioner.la \
-			  $(top_builddir)/lib/common/libcrmcommon.la
+crm_simulate_LDADD	= $(top_builddir)/lib/pacemaker/libpacemaker.la
```
Comment (kgaillot) [inline tools/Makefile.am]: crm_simulate uses all these directly except liblrmd

File: tools/Makefile.am [hunk 7]
```diff
@@ -71,46 +71,28 @@ crm_node_LDADD		= $(top_builddir)/lib/cib/libcib.la		\
...
 crm_diff_LDADD		= $(top_builddir)/lib/common/libcrmcommon.la
 
 crm_mon_SOURCES		= crm_mon.c
-crm_mon_LDADD		= $(top_builddir)/lib/pengine/libpe_status.la	\
-			  $(top_builddir)/lib/fencing/libstonithd.la	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la	\
-			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/common/libcrmcommon.la	\
+crm_mon_LDADD		= $(top_builddir)/lib/fencing/libstonithd.la	\
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la	\
```
Comment (kgaillot) [inline tools/Makefile.am]: crm_mon uses all these directly

File: tools/Makefile.am [hunk 8]
```diff
@@ -71,46 +71,28 @@ crm_node_LDADD		= $(top_builddir)/lib/cib/libcib.la		\
...
 			  $(CURSESLIBS)
 
 # Arguments could be made that this should live in crm/pengine
 crm_verify_SOURCES	= crm_verify.c
-crm_verify_LDADD	= $(top_builddir)/lib/pengine/libpe_status.la 	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la	\
-			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/common/libcrmcommon.la
+crm_verify_LDADD	= $(top_builddir)/lib/pacemaker/libpacemaker.la
```
Comment (kgaillot) [inline tools/Makefile.am]: crm_verify uses all these directly

File: tools/Makefile.am [hunk 9]
```diff
@@ -71,46 +71,28 @@ crm_node_LDADD		= $(top_builddir)/lib/cib/libcib.la		\
...
-			  $(top_builddir)/lib/fencing/libstonithd.la	\
-			  $(top_builddir)/lib/lrmd/liblrmd.la 		\
-			  $(top_builddir)/lib/services/libcrmservice.la \
-			  $(top_builddir)/lib/pengine/libpe_status.la 	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la	\
-			  $(top_builddir)/lib/transition/libtransitioner.la \
-			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/common/libcrmcommon.la
+crm_resource_LDADD	= $(top_builddir)/lib/fencing/libstonithd.la	\
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
Comment (kgaillot) [inline tools/Makefile.am]: crm_resource uses all these directly

File: tools/Makefile.am [hunk 10]
```diff
@@ -126,10 +108,7 @@ attrd_updater_LDADD	= $(top_builddir)/lib/common/libcrmcommon.la
 
 crm_ticket_SOURCES	= crm_ticket.c
 crm_ticket_LDADD	= $(top_builddir)/lib/pengine/libpe_rules.la	\
-			  $(top_builddir)/lib/pengine/libpe_status.la	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la	\
-			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/common/libcrmcommon.la
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
Comment (kgaillot) [inline tools/Makefile.am]: crm_ticket uses all these directly

File: daemons/fenced/pacemaker-fenced.c (hunk 1 above)
Comment (jnpkrn) [reply daemons/fenced/pacemaker-fenced.c]: I didn't mean to put `libpacemaker` and static library into one equation,
it was meant to address some parts of what would remain "private yet still
shared libs" (such as ncurses-based output formatters).

File: daemons/execd/Makefile.am [hunk 11]
```diff
@@ -38,12 +38,9 @@ pacemaker_remoted_LDADD		= $(pacemaker_execd_LDADD) \
...
 				  remoted_tls.c remoted_proxy.c
 
-cts_exec_helper_LDADD	= $(top_builddir)/lib/common/libcrmcommon.la	\
-			  $(top_builddir)/lib/lrmd/liblrmd.la		\
+cts_exec_helper_LDADD	= $(top_builddir)/lib/lrmd/liblrmd.la		\
 			  $(top_builddir)/lib/cib/libcib.la		\
-			  $(top_builddir)/lib/services/libcrmservice.la	\
-			  $(top_builddir)/lib/pengine/libpe_status.la	\
-			  $(top_builddir)/lib/internal/libpacemaker-internal.la
+			  $(top_builddir)/lib/pengine/libpe_status.la
```
[the remark this answers is not in the document]
Comment (jnpkrn) [reply daemons/execd/Makefile.am]: > (This is an example of where consistently using unique prefixes per
> library would make something easier ...)

Cannot agree more, whatever will help understanding "where we stand now"
not a head-banging exercise.  (stating obvious, understanding that question
is a prerequisite to answering to "how to get where we want to be)

**Comment by philbertb** [comment]
Hello Forgive me for posting this issue here as i am a new to github and to this channel.

Can you help me with this error:

virtual_ip	(ocf::heartbeat:IPaddr2):	Stopped

also,

I am getting this to much stonith-enabled due to mispelled config in it s property. I need help in getting rid of this unwanted.

property cib-bootstrap-options: \
	have-watchdog=false \
	dc-version=1.1.14-70404b0 \
	cluster-infrastructure=corosync \
	cluster-name=debian \
	stonith-enabled=false \
	no-quorum-policy=ignore \
	last-lrm-refresh=1554385737 \
	stonith-enabled-false \
	stonith-enabled-false \
	stonith_enabled=true \
	stonith-enabled-false \
	stonith-enabled-false \
	--delete


Thank you

Philbert

**Comment by jnpkrn** [comment]
@philbertb 
It may not be apparent, but this is not the right place to ask for
a community served troubleshooting.  A good start towards resolving
problems with the pacemaker/corosycn cluster stack is
https://clusterlabs.org/help.html, i.e.,

* asking at the dedicated ML (make sure you've at least tried
  searching for the answer beforehand):
  https://lists.clusterlabs.org/mailman/listinfo/users

* for pacemaker issues in particular, you are free to make an
  account at https://bugs.clusterlabs.org/ and log your problem
  there

Regarding your problem at hand, not sure what's the snippet you've
provided exactly happening to be (`crm` specific output, I assume),
but you may look around `crm` (or `pcs`) commands as to how to drop
bogus, malformed `stonith-enabled=<false|true>` entries within the
cluster properties.  Hopefully that's a good start and thanks for
understanding that selecting the right venues to discuss your seemingly
relevant troubles is crucial.  Also, would be wise for you to update,
since pacemaker 1.1.14 is quite ancient by today's standard, and a lot
might have been fixed that you can be hitting otherwise.

**Comment by kgaillot** [comment]
Thanks!
